tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -989,9 +990,18 @@ static const ArTypeObjectKind g_NullTT[] = {AR_TOBJ_VOID, AR_TOBJ_UNKNOWN};
 
 static const ArTypeObjectKind g_ArrayTT[] = {AR_TOBJ_ARRAY, AR_TOBJ_UNKNOWN};
 
+static const ArTypeObjectKind g_ScalarOnlyTT[] = {
+    AR_TOBJ_SCALAR, AR_TOBJ_NOCAST, AR_TOBJ_UNKNOWN};
+
+static const ArTypeObjectKind g_VectorOnlyTT[] = {
+    AR_TOBJ_VECTOR, AR_TOBJ_NOCAST, AR_TOBJ_UNKNOWN};
+
+static const ArTypeObjectKind g_MatrixOnlyTT[] = {
+    AR_TOBJ_MATRIX, AR_TOBJ_NOCAST, AR_TOBJ_UNKNOWN};
+
 const ArTypeObjectKind *g_LegalIntrinsicTemplates[] = {
-    g_NullTT, g_ScalarTT, g_VectorTT, g_MatrixTT,
-    g_AnyTT,  g_ObjectTT, g_ArrayTT,
+    g_NullTT,   g_ScalarTT, g_VectorTT,     g_MatrixTT,     g_AnyTT,
+    g_ObjectTT, g_ArrayTT,  g_ScalarOnlyTT, g_VectorOnlyTT, g_MatrixOnlyTT,
```

Comment:
Automatically generated code for intrinsics will assign indexes into this array that will use the `Only` variants that have a `NOCAST` entry which will stop iterating through possibilities in a way that makes it clear that it failed due to requiring a cast where none is allowed.